### PR TITLE
Improve UI for creating and editing plugins.

### LINKED ISSUE #41736: UI for Project settings ->Plugins is confusing
**Godot version: 3.2 and all later**

**OS/device including version: Windows**

**Issue description:**

The Plugins UI in Project settings can be confusing.  You must make sure to change the addon directory in the popup that generates the plugin, otherwise it will silently delete the other plugin.

**Steps to reproduce:**
 Go to Project Settings->Plugins->Create. After creating the plugin, go back and make another one. It will overwrite the previously created plugins.
See here on reddit for similar: https://www.reddit.com/r/godot/comments/igbv66/do_plugins_work_on_godot_mono_properly/

**Minimal reproduction project: n/a**

### PR DESCRIPTION
**Pull request by EricEzaM**
Commit 1: Moves `set_as_minsize()` from `Popup` to `Window`. This method only invokes methods in `Window`, and it is a nice helper to have on `Window` instead of just `Popup`, ~and I have used it in Commit 2.~

Commit 2: Improves plugin creation dialog by reporting validation errors in the form. Closes #41736

plugin_creation_dialog_final

### CODE REVIEW

**Comment by EricEzaM** [comment]
Oh I also changed `tool` to `@tool` to match GDScript 2.0 convention. I know this section of the file is a `TODO` for a rework but this is just so at least it doesn't throw error when you made a plugin.

**Comment by jkb0o** [comment]
IMO, changing the size of the popup because of the number of warnings isn't leading to good UX. It would be cool if the box with warnings keeps constant minimum size and just does not grow. I guess, as far as all warnings should be fixed to create the plugin, it's ok if we omit messages that don't fit.

**Comment by EricEzaM** [comment]
Not sure if this is really cherry pickable for 3.2 since in 3.2 Popup does not inherit from Window.

@jkb0o I made it so because to me it seems strange to have a large void of nothing when there are no errors.

**Comment by jkb0o** [comment]
When large void of nothing fights unnatural twitching, I bet on the first one. But it is not the only opinion. Anyway, it is nice to see warnings reported this way (inside dialog).

**Comment by Calinou** [comment]
I agree it would be better to reserve some space for errors too. We can use the same pattern as the script creation dialog where a Panel is used to contain error messages (with a darker background).

**Comment by MarianoGnu** [comment]
@EricEzaM instead of showing errors on the bottom you could show error/warning icons next to the field with problems, and click or hover it could display the actual error, something like this;

image

**Comment by EricEzaM** [comment]
Here is what I have at the moment, however I am considering (and have requested feedback regarding) abstracting this feature and making something like a `ValidatedLineEdit` control which is only used in the editor.

validation

This could be useful to replace similar instances of this sort of UI, as below, and also make form-like user interface elements easier to implement in the future.
image
image

**Comment by EricEzaM** [comment]
After further experimentation I don't think it is the best option to make some general `Validated` control. At least not on a small scale. So for now I think this should be implemented on an ad-hoc basis. Below is the final product. If this is merged I will also port this functionality to the script creation dialog and the project settings advanced menu, if desired.

proj_settings_done

**Comment by Calinou** [comment]
@EricEzaM There's the `StatusSuccess` icon you can use instead of the CheckBox node icon.

**Comment by EricEzaM** [comment]
@Calinou don't know how I missed that StatusSuccess icon lol

plugin_creation_dialog_final

**Comment by Jummit** [comment]
Note that some of this functionality would be obsolete if #36677 is merged.

**Comment by Calinou** [comment]
I thought #36677 was merged already :slightly_frowning_face:

**Comment by EricEzaM** [comment]
@Jummit It wouldn't be completely obsolete. Just because the fields are optional does not mean that they do not require validation if they _are_ filled in. If the user decides to set those fields then the values still need to be valid.

Additionally, if the plugin subfolder was still to match an existing folder, then the existing folder contents would still be overwriten which is undesirable.

If that pr was merged then the logic would just need to change so that blank script name = valid. If non blank, the validation would still need to occur. Everything else would stay as far as I can see.

**Review by Calinou — COMMENTED** [review]
Should be good to merge after rebasing. I'll salvage changes from #36677 that are still relevant once this PR is merged.

**Note to maintainers:** If there are plans to cherry-pick this to the `3.x` branch, make sure not to cherry-pick the `@tool` change :slightly_smiling_face:

**Review by Calinou — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
